sdk/formrecognizer/Azure.AI.FormRecognizer/src/StreamExtensions.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using Azure.AI.FormRecognizer.Models;
+
+namespace Azure.AI.FormRecognizer
+{
+    internal static class StreamExtensions
+    {
+        /// <summary>The set of bytes expected to be present at the start of PDF files.</summary>
+        private static byte[] PdfHeader = new byte[] { 0x25, 0x50, 0x44, 0x46 };
+
+        /// <summary>The set of bytes expected to be present at the start of PNG files.</summary>
+        private static byte[] PngHeader = new byte[] { 0x89, 0x50, 0x4E, 0x47 };
+
+        /// <summary>The set of bytes expected to be present at the start of JPEG files.</summary>
+        private static byte[] JpegHeader = new byte[] { 0xFF, 0xD8 };
+
+        /// <summary>The set of bytes expected to be present at the start of TIFF (little-endian) files.</summary>
+        private static byte[] TiffLeHeader = new byte[] { 0x49, 0x49, 0x2A, 0x00 };
+
+        /// <summary>The set of bytes expected to be present at the start of TIFF (big-endian) files.</summary>
+        private static byte[] TiffBeHeader = new byte[] { 0x4D, 0x4D, 0x00, 0x2A };
+
+        /// <summary>
+        /// Attemps to detect the <see cref="ContentType"/> of a stream of bytes. The algorithm searches through
+        /// the first set of bytes in the stream and compares it to well-known file signatures.
+        /// </summary>
+        /// <param name="stream">The stream to which the content type detection attempt will be performed.</param>
+        /// <param name="contentType">If the detection is successful, outputs the detected content type. Otherwise, <c>default</c>.</param>
+        /// <returns><c>true</c> if the detection was successful. Otherwise, <c>false</c>.</returns>
+        /// <exception cref="NotSupportedException">Happens when <paramref name="stream"/> is not seekable or readable.</exception>
+        public static bool TryGetContentType(this Stream stream, out ContentType contentType)
+        {
+            var originalPosition = stream.Position;
+            var bytesCount = Math.Min(stream.Length, PdfHeader.Length);
+
+            bool hasFoundType = false;
+            contentType = default;
+
+            bool couldBePdf = true;
+            bool couldBePng = true;
+            bool couldBeJpeg = true;
+            bool couldBeTiffLe = true;
+            bool couldBeTiffBe = true;
+
+            Func<byte[], int, bool> isAtEnd = (array, index) => index == array.Length - 1;
+            Func<byte[], int, byte, bool> hasByteAt = (array, index, expected) => index < array.Length && array[index] == expected;
+
+            for (int i = 0; i < bytesCount; i++)
```

Comment:
It looks like all headers are different on their first byte, so if first byte is matched, we can check just that header.

Reply:
Something like:
```cs
var firstByte = stream.ReadByte();
if (IsHeader(stream, firstByte, PdfHeader)) 
{
    contentType = ContentType.Pdf;
}
else if (IsHeader(stream, firstByte, PngHeader)) 
{
    contentType = ContentType.Png;
}
else if (IsHeader(stream, firstByte, JpegHeader)) 
{
    contentType = ContentType.Jpeg;
}
else if (IsHeader(stream, firstByte, TiffBeHeader) || IsHeader(stream, firstByte, TiffLeHeader)) 
{
    contentType = ContentType.Tiff;
}
else
{
    stream.Position = originalPosition;
    return false;
}

stream.Position = originalPosition;
return true;
```
where `IsHeader` looks like this:
```cs
private static bool IsHeader(Stream stream, byte firstByte, byte[] header)
{
    if (stream.Length - stream.Position >= header.Length || firstByte != header[0]) 
    {
        return false;
    }
    for (var i = 1; i < header.Length; i++) 
    {
        var currentByte = stream.ReadByte();
        if (currentByte != header[i]) 
        {
            return false;
        }
    }
    return true;
}
```

Reply:
Or, since `CanSeek == true`, it can be even simpler:
```cs
if (IsHeader(stream, PdfHeader)) 
{
    contentType = ContentType.Pdf;
}
else if (IsHeader(stream, PngHeader)) 
{
    contentType = ContentType.Png;
}
else if (IsHeader(stream, JpegHeader)) 
{
    contentType = ContentType.Jpeg;
}
else if (IsHeader(stream, TiffBeHeader) || IsHeader(stream, TiffLeHeader)) 
{
    contentType = ContentType.Tiff;
}
else
{
    return false;
}

return true;
```
and
```cs
private static bool IsHeader(Stream stream, byte[] header)
{
    if (stream.Length - stream.Position > header.Length) 
    {
        return false;
    }

    var originalPosition = stream.Position;
    foreach (var headerByte in header) 
    {
        var currentByte = stream.ReadByte();
        if (currentByte != headerByte) 
        {
            stream.Position = originalPosition;
            return false;
        }
    }
    stream.Position = originalPosition;
    return true;
}
```

Reply:
The second approach is way simpler (it's something Anne suggested as well). I thought about doing something like that at first, but gave up because of the extra read operations. I'm just paying too much attention to details, though. We have whole image files in the stream, so this trifling performance difference is easily compensated by the simplicity of the code.

By the way, subtracting the original `stream.Position` from `stream.Length` was a smart move. I almost introduced a bug there.